### ISSUE TITLE
Identifier is not defined/Cannot open stddef.h (dependency of stdlib.h)

### ISSUE THREAD

**Issue by namarino41** [microsoft/vscode-cpptools#1202]
- VSCode Version: 1.17
- OS Version: Ubuntu 16.04

Steps to Reproduce:

1. Write simple c program.
2. #include <stdlib.h>

Extensions installed: C/C++ extension by Microsoft

When I include stdlib.h, it gets a green squiggly saying that it can not load one of it's dependencies and to edit the include path. When I do that and add /usr/include/linux, that squiggly line goes away, but then I get error messages on some of my variables that are from other header files. None of this effects the way the program runs. Everything compiles and runs just fine.
![image](https://user-images.githubusercontent.com/31663851/32360912-bf817d9e-c030-11e7-90d4-8fb1b6ca8ad6.png)

![image](https://user-images.githubusercontent.com/31663851/32360914-c6a22d94-c030-11e7-953e-6bb41f6411ca.png)

**Comment by sean-mcmanus** [comment]
What is your includePath setting? I get this error when I have "/usr/include/linux" on my path, but removing it fixes the issue (along with time_t being undefined when using time.h).

**Comment by namarino41** [comment]
```
{
            "name": "Linux",
            "includePath": [
                "/usr/include/x86_64-linux-gnu/c++/5",
                "/usr/include/c++/5",
                "/usr/local/include",
                "/usr/include/x86_64-linux-gnu",
                "/usr/include",
                "/usr/include/linux",
                "${workspaceRoot}"                
            ],
            "defines": [],
            "intelliSenseMode": "clang-x64",
            "browse": {
                "path": [
                    "/usr/include/x86_64-linux-gnu/c++/5",
                    "/usr/include/c++/5",
                    "/usr/local/include",
                    "/usr/include/x86_64-linux-gnu",
                    "/usr/include",
                    "${workspaceRoot}"
                ],
                "limitSymbolsToIncludedHeaders": false,
                "databaseFilename": ""
            }
        },
```

**Comment by sean-mcmanus** [comment]
Can you remove `/usr/include/linux` from the includePath or is it required? Do your include paths and ordering match what's given by `clang -v -E -x c++ - < /dev/null` (with clang replaced by gcc or g++ if you're using those)?

**Comment by namarino41** [comment]
Well if I remove that path, I get a green squiggly under include <stdlib.h> that says it cannot open source file "stddef.h" (dependency of "stdlib.h"). 
Here is the output of that `clang` command, but using `gcc`:
```
Using built-in specs.
COLLECT_GCC=gcc
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 5.4.0-6ubuntu1~16.04.5' --with-bugurl=file:///usr/share/doc/gcc-5/README.Bugs --enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-5 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-5-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-5-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-5-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.5) 
COLLECT_GCC_OPTIONS='-v' '-E' '-mtune=generic' '-march=x86-64'
 /usr/lib/gcc/x86_64-linux-gnu/5/cc1plus -E -quiet -v -imultiarch x86_64-linux-gnu -D_GNU_SOURCE - -mtune=generic -march=x86-64 -fstack-protector-strong -Wformat -Wformat-security
ignoring duplicate directory "/usr/include/x86_64-linux-gnu/c++/5"
ignoring nonexistent directory "/usr/local/include/x86_64-linux-gnu"
ignoring nonexistent directory "/usr/lib/gcc/x86_64-linux-gnu/5/../../../../x86_64-linux-gnu/include"
#include "..." search starts here:
#include <...> search starts here:
 /usr/include/c++/5
 /usr/include/x86_64-linux-gnu/c++/5
 /usr/include/c++/5/backward
 /usr/lib/gcc/x86_64-linux-gnu/5/include
 /usr/local/include
 /usr/lib/gcc/x86_64-linux-gnu/5/include-fixed
 /usr/include/x86_64-linux-gnu
 /usr/include
End of search list.
# 1 "<stdin>"
# 1 "<built-in>"
# 1 "<command-line>"
# 1 "/usr/include/stdc-predef.h" 1 3 4
# 1 "<command-line>" 2
# 1 "<stdin>"
COMPILER_PATH=/usr/lib/gcc/x86_64-linux-gnu/5/:/usr/lib/gcc/x86_64-linux-gnu/5/:/usr/lib/gcc/x86_64-linux-gnu/:/usr/lib/gcc/x86_64-linux-gnu/5/:/usr/lib/gcc/x86_64-linux-gnu/
LIBRARY_PATH=/usr/lib/gcc/x86_64-linux-gnu/5/:/usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/:/usr/lib/gcc/x86_64-linux-gnu/5/../../../../lib/:/lib/x86_64-linux-gnu/:/lib/../lib/:/usr/lib/x86_64-linux-gnu/:/usr/lib/../lib/:/usr/lib/gcc/x86_64-linux-gnu/5/../../../:/lib/:/usr/lib/
COLLECT_GCC_OPTIONS='-v' '-E' '-mtune=generic' '-march=x86-64'
```

**Comment by sean-mcmanus** [comment]
Adding "/usr/lib/gcc/x86_64-linux-gnu/5/include" should fix your issue (and removing the linux folder), but ideally you should use the complete list as shown (from "/usr/include/c++/5" to "/usr/include") or you could hit other issues. I think our next update will automatically try to use the more accurate defaults from the compiler.

**Comment by namarino41** [comment]
That fixed it. I'l go ahead and add that the rest of those paths to my paths file. Thank you so much for your help, I really appreciate it.

**Comment by FreddieChopin** [comment]
I'm getting the same error. With the `compile_commands.json` generated by CMake (project is building fine) I get this error when opening any file in the project. So I've edited down my file down to a single entry, with almost everything removed:

```
[
{
  "directory": "/home/freddie/Elektronika/ARM/Projects/STM32F4-mono-graphic-keypad",
  "command": "g++ -IWatchdogTimer",
  "file": "/home/freddie/Elektronika/ARM/Projects/STM32F4-mono-graphic-keypad/watchdogTimerManagerInstance.cpp"
}
]
```

I reset IntelliSense database, reopen the file and see `Failed to query default include paths for use with compile_commands.json` again, included file which is in the `WatchdogTimer` folder is not found.

I'm looking forward to next release with the new options, but I'm afraid these won't be enough for embedded projects which use cross-compilers which also need special flags. In my case I get different internal compiler defines depending on the flags I used. For example with `arm-none-eabi-gcc -dM -E - < /dev/null` I get slighly different set than with `arm-none-eabi-gcc -dM -E -mcpu=cortex-m4 -mthumb -mfloat-abi=hard -mfpu=fpv4-sp-d16 - < /dev/null` and this is slightly changed when using `arm-none-eabi-gcc -dM -E -mcpu=cortex-m4 -mthumb - < /dev/null`. Will that be possible to put the path to compiler _WITH_ the required flags in the new `compilerPath` property? If not, would it be possible to change it before the final release?
